### DISCUSSION TITLE
Output shape of allgather(): Horovod vs Pytorch

### DISCUSSION THREAD

**Discussion by Alam45** [horovod/horovod#3167] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
The output of horovod.torch.allgather(tensor, name=None) is a tensor of concatenation of tensor on different rank processes. While the output of torch.distributed.all_gather(tensor_list, tensor) is a tensor list. 

Pytorch output (Two GPUs):
before all_gather(): Input tensors
    #tensor([1, 2]) # Rank 0
    #tensor([3, 4]) # Rank 1
After all_gather(): Output tensor list
    #[tensor([1, 2]), tensor([3, 4])] # Rank 0
    #[tensor([1, 2]), tensor([3, 4])] # Rank 1

Horovod output (Two GPUs):
before all_gather(): Input tensors
    #tensor([1, 2]) # Rank 0
    #tensor([3, 4]) # Rank 1
After all_gather(): Output tensor
    #tensor([1, 2, 3, 4]) # Rank 0
    #tensor([1, 2, 3, 4]) # Rank 1

How can I get the output as in pytorch using horovod.torch.allgather(tensor, name=None)?

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by maxhgerlach** [comment, marked as answer]
@Mahbubalam4545, have you tried passing two-dimensional tensors into `hvd.allgather()`? E.g., they would be shaped like `[[1,2]]` and `[[3,4]]`.

**Reply by Alam45** [reply]
I've tried passing two-dimensional tensors into hvd.allgather(). However, it didn't wrok. 
Input:
tensor([[7, 8]], device='cuda:3')
tensor([[3, 4]], device='cuda:1')
tensor([[5, 6]], device='cuda:2')
tensor([[1, 2]], device='cuda:0')

Output:
tensor([[1, 2], [3, 4], [5, 6], [7, 8]], device='cuda:0')
tensor([[1, 2], [3, 4], [5, 6], [7, 8]], device='cuda:1')
tensor([[1, 2], [3, 4], [5, 6], [7, 8]], device='cuda:3')
tensor([[1, 2], [3, 4], [5, 6], [7, 8]], device='cuda:2')

**Reply by maxhgerlach** [reply]
I don't understand what you are missing? If you absolutely need a list of tensors instead of a large multidimensional tensor, just split it up?

**Reply by Alam45** [reply]
Yes, I have split it in my code before this PR. My argument is why Horovod is merging the tensors gathered from different processes? and what are the benefits? If in a situation when the tensors are of varied sizes or when we don't know the exact size of the tensors on each of the processes, in these situations the allgather() would be useless. In short, I would request you to investigate the intuitions behind the design of Pytorch's all_gather().

**Reply by maxhgerlach** [reply]
Allgather's main purpose in Horovod is to aggregate sparse gradients.

You could also look at `hvd.allgather_object` if you want a convenient function that gives you a Python list.

**Reply by Alam45** [reply]
I'm using horovod 0.19.0. does it have `allgather_object`? I'm getting error:
 "AttributeError: module  `horovod.torch` has no attribute `allgather_object` "
Kindly, advise me, Which version should I use?

**Reply by maxhgerlach** [reply]
you should definitely upgrade to the latest release :)

**Reply by bryant1410** [reply]
Does `allgather_object` respect the gradients like `dist.nn.all_gether` does?
